### PR TITLE
MLPF v2 on 12_1_0_pre3

### PR DESCRIPTION
**Pull request by jpata**
- based on the MLPF model benchmark training at https://github.com/jpata/particleflow/pull/84, https://github.com/jpata/particleflow/pull/92 and https://github.com/jpata/particleflow/pull/90
- The final training was carried out by Eric
- switches TF->ONNX (CPU) for inference
- improves the integration by correctly setting the references of the PFCandidates to the originator particles. Puppi and CHS now work.
- data files in https://github.com/jpata/RecoParticleFlow-PFProducer/tree/mlpf_v2_CMSSW_12_1_0_pre3_updated
- mlpf workflows are now 11834.13 (TTbar) and 11843.13 (QCD)
- slides for this PR are here: https://indico.cern.ch/event/1093938/#6-mlpf-in-cms

### CODE REVIEW

File: RecoParticleFlow/PFProducer/interface/MLPFModel.h [hunk 1]
```diff
@@ -7,31 +7,40 @@
...
-  static constexpr int NUM_MAX_ELEMENTS_BATCH = 20000;
-  static constexpr int LSH_BIN_SIZE = 100;
+  static constexpr int LSH_BIN_SIZE = 64;
+  static constexpr int NUM_MAX_ELEMENTS_BATCH = 200 * LSH_BIN_SIZE;
 
-  //In CPU mode, we only want to evaluate each event separately
+  //In CPU mode, we want to evaluate each event separately
   static constexpr int BATCH_SIZE = 1;
 
   //The model has 12 outputs for each particle:
```
Comment (jmduarte) [inline RecoParticleFlow/PFProducer/interface/MLPFModel.h:20]: 14 outputs

File: RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc [hunk 2]
```diff
@@ -4,135 +4,165 @@
...
 }
 
-void MLPFProducer::globalEndJob(MLPFCache* cache) { delete cache->graph_def; }
+void MLPFProducer::globalEndJob(const ONNXRuntime* cache) {}
 
 void MLPFProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
   edm::ParameterSetDescription desc;
   desc.add<edm::InputTag>("src", edm::InputTag("particleFlowBlock"));
-  desc.add<std::string>("model_path", "RecoParticleFlow/PFProducer/data/mlpf/mlpf_2020_11_04.pb");
+  desc.add<edm::FileInPath>("model_path", edm::FileInPath("RecoParticleFlow/PFProducer/data/mlpf/mlpf_2021_10_27__all_data_cms-best-of-asha-scikit_20211026_042043_178263.workergpu010.onnx"));
```
Comment (jmduarte) [inline RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc:165]: may want to have a subdirectory or a differently named directory, e.g. `mlpf_acat2021`?

File: RecoParticleFlow/PFProducer/interface/MLPFModel.h (hunk 1 above)
Comment (jpata) [reply RecoParticleFlow/PFProducer/interface/MLPFModel.h:20]: thanks, will fix

File: RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc (hunk 2 above)
Comment (jpata) [reply RecoParticleFlow/PFProducer/plugins/MLPFProducer.cc:165]: I think I would just keep it as-is for simplicity.

**Comment by jpata** [comment]
This was merged in the official CMSSW, so I'm closing it here.

**Pull request closed without merging**